tests/instrument/expr_match.ml.reference

```diff
@@ -99,3 +99,174 @@ let f x =
         (match x with
          | Foo  -> (___bisect_mark___expr_match 36; print_endline "foobar")
          | Bar  -> (___bisect_mark___expr_match 37; print_endline "barbar"))))
+let f x =
+  ___bisect_mark___expr_match 43;
+  (match x with
+   | Foo |Bar  as ___bisect_matched_value___ ->
+       ((((match ___bisect_matched_value___ with
+           | Foo  -> (___bisect_mark___expr_match 41; ())
+           | Bar  -> (___bisect_mark___expr_match 42; ())
+           | _ -> ()))[@ocaml.warning "-8-11"]);
+        print_endline "foo"))
```

Comment:
This is a good counterexample, that I can think can lead to better ones. It weirdly reminds of how side effect expressions are not allowed in `Erlang` guards, which at the time I thought was impure. 

I'd include the first 2 paragraphs of your explanation in the doc.